corehq/apps/couch_sql_migration/couchsqlmigration.py

```diff
@@ -391,6 +394,100 @@ def _process_missing_forms(self):
                     if migrated % 100 == 0:
                         log.info("migrated %s previously missed forms", migrated)
         log.info("finished migrating %s previously missed forms", migrated)
+        self._process_missing_case_references()
+
+    def _process_missing_case_references(self):
+        """Extract forms from case diffs and process missing elements"""
+        def iter_form_ids(jdiff, kind):
+            old_value = jdiff.old_value
+            if old_value is MISSING or not old_value:
+                return
+            if kind == "CommCareCase":
+                if isinstance(old_value, dict) and "forms" in old_value:
+                    for form_id, status in old_value["forms"].items():
+                        if status != MISSING_BLOB_PRESENT:
+                            yield form_id
+                elif jdiff.diff_type == 'set_mismatch' and jdiff.path[0] == 'xform_ids':
+                    yield from old_value.split(",")
+            elif (
+                kind == "stock state"
+                and isinstance(old_value, dict)
+                and "ledger" in old_value
+                and old_value.get("form_state") != MISSING_BLOB_PRESENT
+            ):
+                yield old_value["ledger"]["last_modified_form_id"]
+
+        from .missingdocs import MissingIds
+        iter_blob_forms = _get_missing_blob_present_form_loader(self.domain)
+        drop_sql_ids = MissingIds.forms(self.statedb).drop_sql_ids
+        for diff in _iter_case_diffs(self.statedb, self.stopper):
+            case_id = diff.doc_id
+            for form in iter_blob_forms(diff):
+                log.info("migrating form %s received on %s from case %s",
+                    form.form_id, form.received_on, case_id)
+                self._migrate_form(form, get_case_ids(form))
+            form_ids = set(iter_form_ids(diff.json_diff, diff.kind))
+            if not form_ids:
+                continue
+            missing_ids = set(drop_sql_ids(form_ids))
+            for form_id in missing_ids:
+                form = XFormInstance.get(form_id)
+                self._migrate_form(form, get_case_ids(form))
+                log.info("migrated missing form %s from case %s", form_id, case_id)
+            self._migrate_missing_cases_and_ledgers(form_ids - missing_ids)
+
+    def _migrate_missing_cases_and_ledgers(self, form_ids):
+        """Update cases and ledgers for forms that have already been migrated
+
+        :returns: a set of form ids that had cases or ledgers to migrate.
+        """
+        migrated_ids = set()
+        for form_id in form_ids:
+            saved = self._save_missing_cases_and_ledgers(form_id)
+            if saved:
+                migrated_ids.add(form_id)
+        return migrated_ids
+
+    def _save_missing_cases_and_ledgers(self, form_id):
+        def iter_missing_ledgers(stock_result):
```

Comment:
> If there is a concern that the `LedgerValue` might exist but may be missing some transactions

Yes, this is the situation I'm trying to resolve here.

> then you should use `transaction.natural_key()` to look for missing ones (not `transaction.ledger_reference`) since the natural key includes `form_id`.

The result of `LedgerAccessorSQL.get_ledger_transactions_for_form(form_id, case_ids)` only includes transactions relevant to *one form*, and the ledger values I'm checking all came out of that form. It should yield the same result whether this code matches on `ledger_reference` or `natural_key()`, right?

I am curious if it is possible for a single form to have more than one ledger transaction for a given ledger value? I think this is not supported, based on the data model keys, but I may have overlooked something.

> If it turns out that a transaction is missing for an existing `LedgerValue` then you'd also need to rebuild the `LedgerValue`.

I think this is at least part of what is missing. Hopefully it will resolve the issue outlined in https://github.com/dimagi/commcare-hq/pull/26768/files/082acc9bd7121a1047ef35a4c16aea0fdbd7698a#r388588292

Reply:
> It should yield the same result whether this code matches on ledger_reference or natural_key(), right?

True, in this case it would since all the form_id values would be the same

>I am curious if it is possible for a single form to have more than one ledger transaction for a given ledger value? I think this is not supported, based on the data model keys, but I may have overlooked something.

Interestingly you can. There's nothing to prevent that in the code. I don't think that's an intentional decision though. Would be interesting to see if that case exists in the wild.